include/vapor/ArbitrarilyOrientedRegularGrid.h

```diff
@@ -0,0 +1,114 @@
+//! 2D RegularGrid that are outside of the valid extents will be assigned as missingValues.  This Grid class
+//! will always be valid, even if the the plane's description lies outside of the box.  In this case, the
+//! plane will contain all missing values.
+//!
+//! Note: The boxMin and boxMax values within the planeDescription should come from
+//! the current RenderParams' Box class.  The domainMin and domainMax values should
+//! come from the DataMgr::GetVariableExtents function.  Taking these values from
+//! the given 3D grid can produce erroneous results when the 3D grid's region is reduced.
+//!
+//! \param[in] grid3d A variable's 3D grid to be sampled along a plane, to generate a 2D grid object
+//! \param[in] description A set of std::vector<dobule> values that define the origin, rotation, and extents of the returned 2D grid
+//! \param[in] dims The sampling rate that the new grid will be defined upon, along its X and Y axes
+//! \param[out] data An array of sideSize*sideSize values that contain floating point values of the sampled data.
+//!
+
+class VDF_API ArbitrarilyOrientedRegularGrid : public RegularGrid {
+public:
+    ArbitrarilyOrientedRegularGrid(
+        const VAPoR::Grid* grid3d,
+        planeDescription& pd,
+        const DimsType& dims,
+        std::shared_ptr<float>& blks
+    );
+
+    ArbitrarilyOrientedRegularGrid() = default;
+    virtual ~ArbitrarilyOrientedRegularGrid() = default;
+
+    static std::string GetClassType() { return ("ArbitrarilyOrientedRegular"); }
+    std::string        GetType() const override { return (GetClassType()); }
+
+    //! \copydoc Grid::GetUserCoordinates()
+    //
+    virtual void GetUserCoordinates(const DimsType &indices, CoordType &coords) const override;
+    virtual void GetUserCoordinates (const size_t indices[], double coords[]) = delete;
+    virtual void GetUserCoordinates (const std::vector< size_t > &indices, std::vector< double > &coords) = delete;
+    virtual void GetUserCoordinates (size_t i, double &x, double &y, double &z) = delete;
+    virtual void GetUserCoordinates (size_t i, size_t j, double &x, double &y, double &z) = delete;
+    virtual void GetUserCoordinates (size_t i, size_t j, size_t k, double &x, double &y, double &z) = delete;
+
+    //! Retrieve the winding order that pertains to the generated array of data, if it is
+    //! to be drawn as a texture as is done in the SliceRenderer's vertex buffer.  The winding order is a series
+    //! of six 3D vertices that describe two triangles that define how the texture is drawn in space.
+    void GetWindingOrder( std::vector<double>& windingOrder ) const;
+
+    //! Retrieve the 3D coordinates of the corners of the sampled data.  This is a set of four 3D coordinates,
+    //! used to configure the SliceRenderer's texture coordinate buffer.
+    void Get3DRectangle( std::vector<double>& windingOrder ) const;
```

Comment:
John and I talked about Get3DRectangle() and GetWindingOrder() on Friday and we decided to these functions from the VDC library.  These values are now in the SliceRenderer.  I think there is going to be more discussion on this matter going forward.

 PS - This function takes a non-const parameter, which implies that it changes.  Isn't it more efficient to take a value by reference when it's getting modified?  This is our practice with strings.  Is it not so with vectors?

Reply:
In this case, it would not be more efficient to pass by reference. In the case of strings it is also not more efficient. The issue with mutable pass by reference is you have no way of knowing from syntax that the passed in value is an output in the calling code.